piptools/scripts/compile.py

```diff
@@ -42,10 +45,44 @@
     "pyproject.toml",
     "setup.cfg",
 )
+ALL_BUILD_DISTRIBUTIONS: tuple[Literal["sdist", "wheel", "editable"], ...] = (
+    "editable",
+    "sdist",
+    "wheel",
+)
+DEFAULT_REQUIREMENTS_FILE = "requirements.in"
 DEFAULT_REQUIREMENTS_OUTPUT_FILE = "requirements.txt"
 METADATA_FILENAMES = frozenset({"setup.py", "setup.cfg", "pyproject.toml"})
 
 
+def _build_requirements(
+    src_dir: str, src_file: str, distributions: Iterable[str], package_name: str
+) -> list[InstallRequirement]:
+    assert not isinstance(distributions, str)
+    result = collections.defaultdict(set)
+
+    with IsolatedEnvBuilder() as env:
+        builder = ProjectBuilder(
```

Comment:
Could you help me understand why we don't use `build.utils.project_wheel_metadata()` directly?

Reply:
@atugushev I have a hunch this might be because it needs to extract separate dependencies from all the PEP 517 hooks. They are called `distributions` here in the source code, which must be misleading. I think we've gotten rid of the naming on the user-facing CLI side, but it's probably still confusing here.
@apljungquist should we rename `distributions` to `dynamic_build_dependency_targets`? Maybe a docstring would help the readers understand what it contains?

`project_wheel_metadata()` gets us wheel metadata, but we don't need all of it — only the build deps for wheels. We also need sdist build deps and editable build deps. `get_requires_for_build()` is what can be called with one of `'sdist'`, `'wheel'` and `'editable'` and what we actually want to retrieve. Using that helper just for wheels would be pointless, and it returns metadata from a `dist-info` folder — I don't think we'd be able to extract what we need even just for wheels.

Reply:
Aw, I see. We need builder instance. Thanks!